honour SOURCE_DATE_EPOCH environment variable

Villemoes opened the pull request:
In the interest of producing more reproducible builds, honour SOURCE_DATE_EPOCH if set.

Commit:
version.py: honour SOURCE_DATE_EPOCH for more reproducible builds

Doing binary comparisons of a root filesystem built on two different
developers' machines, we found a single difference in a python module
installed using setuptools_scm. In order to enhance binary
reproducibility, honour SOURCE_DATE_EPOCH if that environment variable
is set.

See https://reproducible-builds.org/docs/source-date-epoch/ for
background on reproducible builds and the SOURCE_DATE_EPOCH
environment variable.

Commit:
utils.py: remove unused utc singleton

With the previous patch switching to use datetime.utcfromtimestamp()
in version.py, this singleton tzinfo object is no longer used
anywhere.

RonnyPfannschmidt commented on src/setuptools_scm/version.py:
```diff
@@ -136,7 +138,9 @@ def __init__(
         self.node = node
-        self.time = datetime.datetime.now(utc)
+        self.time = datetime.datetime.utcfromtimestamp(
+            int(os.environ.get("SOURCE_DATE_EPOCH", time.time()))
```
i'd like to see a unittest for that

RonnyPfannschmidt reviewed (changes requested):
please excuse the late review,
for personal reason i had to put a lot into my backlog and i'm just now processing

Commit:
testing: use and test SOURCE_DATE_EPOCH handling

Now that SOURCE_DATE_EPOCH is honoured, we can avoid the need to
monkeypatch the time attribute with a MockTime object, while at the
same time testing handling of SOURCE_DATE_EPOCH being set in the
environment. Most tests will actually benefit from knowing exactly
what to expect, so set SOURCE_DATE_EPOCH globally.

There's an existing test in test_git.py which currently manually
produces the yyyymmdd string to compare to - extend that test a bit to
do it both with and without SOURCE_DATE_EPOCH being set, so that we
keep at least one "today's date" test.

RonnyPfannschmidt commented on testing/test_functions.py:
```diff
@@ -36,6 +31,8 @@ def test_next_tag(tag, expected):
 c = Configuration()
 
+# 2009-02-13T23:31:30+00:00
+os.environ["SOURCE_DATE_EPOCH"] = "1234567890"
```
nice idea

please use `monkeypatch.setenv` inside of the test instead of doing a global patching there

else the behavior bleeds to other tests an

Villemoes replied:
yeah, I thought about that. Problem is that, AFAICT, the `meta()` objects in the `VERSIONS` array get created at global scope, so modifying the environment within the test case doesn't affect the `.time` attribute that's already set. That is, if I do
```
 # 2009-02-13T23:31:30+00:00
-os.environ["SOURCE_DATE_EPOCH"] = "1234567890"
+#os.environ["SOURCE_DATE_EPOCH"] = "1234567890"
 VERSIONS = {
     "exact": meta("1.1", distance=None, dirty=False, config=c),
     "zerodistance": meta("1.1", distance=0, dirty=False, config=c),
@@ -60,7 +60,8 @@ VERSIONS = {
         ("distancedirty", "post-release node-and-date", "1.1.post3+d20090213"),
     ],
 )
-def test_format_version(version, scheme, expected):
+def test_format_version(version, monkeypatch, scheme, expected):
+    monkeypatch.setenv("SOURCE_DATE_EPOCH", "1234567890")
     version = VERSIONS[version]
```
on top, the test fails (as I expected) with
```
E       AssertionError: assert '1.1.post3+d20201014' == '1.1.post3+d20090213'
E         - 1.1.post3+d20201014
E         ?               - ^^^
E         + 1.1.post3+d20090213
E         ?              +++  ^
```
I don't really know much about pytest (I can't even get the full test suite to run), so I'm at a loss of how to add a real test.

Villemoes replied:
hmm, conftest.py already sets one env variable globally (which then gets unset in a few tests). Perhaps SOURCE_DATE_EPOCH should be set there as well and used for all the date tests - after all, the `assert today.strftime(".d%Y%m%d") in wd.version in test_git.py has a microscopic chance of a false negative is the test is run very close to (UTC) midnight.

RonnyPfannschmidt replied:
Good idea,

For pedantic we ought to have a single test that checks the no env var case

Villemoes replied:
All good now?

RonnyPfannschmidt replied:
yup, nicely done, thanks
